fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java

```diff
@@ -100,27 +104,172 @@ public MvRewritePreprocessor(ConnectContext connectContext,
-        try (Timer ignored = Tracers.watchScope("preprocessMvs")) {
-            // MV Rewrite will be used when cbo is enabled.
-            if (context.getOptimizerConfig().isRuleBased()) {
-                return;
-            }
+    public OptExpression getLogicalTree() {
+        return logicalTree;
+    }
+
+    class MvWithPlanContext {
+        private final MaterializedView mv;
+        private final MvPlanContext mvPlanContext;
+        public MvWithPlanContext(MaterializedView mv, MvPlanContext mvPlanContext) {
+            this.mv = mv;
+            this.mvPlanContext = mvPlanContext;
+        }
+
+        public MaterializedView getMv() {
+            return mv;
+        }
+
+        public MvPlanContext getMvPlanContext() {
+            return mvPlanContext;
+        }
+    }
+
+    /**
+     * When {@link MvPlanContext} already existed, check the related mvs by using it
+     * in {@link #chooseBestRelatedMVs(Set, Set, OptExpression)}.
+     */
+    class MVWithPlanContextCache {
+        private final boolean isValidPlan;
+        private final int scanOpNumDiff;
+        public MVWithPlanContextCache(boolean isValidPlan, int scanOpNumDiff) {
+            this.isValidPlan = isValidPlan;
+            this.scanOpNumDiff = scanOpNumDiff;
+        }
+
+        public boolean isValidPlan() {
+            return isValidPlan;
+        }
+
+        public int getScanOpNumDiff() {
+            return scanOpNumDiff;
+        }
+    }
+    
+    /**
+     * To avoid MvRewriteProcessor cost too much optimizer time, reduce all related mvs to a limited size.
+     * <h3>Why to Choose The Best Related MVs Strategy</h3>
+     *
+     * <p>Why still to choose limited related mvs from all active mvs?</p>
+     *
+     * <p>1. optimizer time cost. Even there is MVPlanCache to reduce mv optimizer plan time, but it still may cost
+     * too much time for mv preprocessor, because mv optimizer plan costs too much time because of cold start when
+     * MVPlanCache has no cache, or MVPlanCache has exceeded limited capacity which is 1000 by default.</p>
+     *
+     * <p>2. check mv's freshness for each mv will cost too much time if there are too many mvs.</p>
+     *
+     * <h3>How to Choose The Best Related MVs Strategy</h3>
+     * <p>
+     *     Choose the best related mvs from all active mvs as following order:
+     *     1. find the max intersected table num between mv and query which means it's better for rewrite.
+     *     2. find the latest fresh mv which means its freshness is better.
+     * </p>
+     *
+     * <h3>More Information</h3>
+     * <p>
+     *  NOTE: there are still some limitations about this ordering algorithm:
+     *  1. consider repeated tables in one mv later which one table can be used repeatedly in one mv.
+     *  2. consider random factor so can cache and use more mvs.
+     * </p>
+     */
+    static class MVRelatedOrdering implements Comparator<MaterializedView> {
```

Comment:
make MaterializedViewWithOrder immutable
cache the results of getInteractedTablesNum,getBaseTablesNum, getBaseTablesNum use transient field

Reply:
As you suggests, I have added a class to cache the results like this:
```
    public static class MVCorrelation implements Comparable<MVCorrelation> {
        private final MaterializedView mv;
        private final long mvQueryIntersectedTablesNum;
        private final int mvQueryScanOpNumDiff;
        private final long mvRefreshTimestamp;

        public MVCorrelation(MaterializedView mv,
                             long mvQueryIntersectedTablesNum,
                             int mvQueryScanOpNumDiff,
                             long mvRefreshTimestamp) {
            this.mv = mv;
            this.mvQueryIntersectedTablesNum = mvQueryIntersectedTablesNum;
            this.mvQueryScanOpNumDiff = mvQueryScanOpNumDiff;
            this.mvRefreshTimestamp = mvRefreshTimestamp;
        }
}
```